mesonbuild/coredata.py

```diff
@@ -599,7 +599,7 @@ def update_project_options(self, project_options: 'MutableKeyedOptionDictType',
             oldval = self.optstore.get_value_object(key)
             if type(oldval) is not type(value):
                 self.optstore.set_value(key, value.value)
-            elif oldval.choices != value.choices:
+            elif oldval.printable_choices() != value.printable_choices():
```

Comment:
Using `printable_choices()` for this is a bit strange. I think it is done to accommodate for `UserIntegerOption` which AFAICT is the only option type that specializes `printable_choices()` to return something not computed from its `choices` member.

Reply:
I haven't checked if something else would break, but having `UserIntegerOption` return a `slice` object for its `choice` property would solve this maybe a bit more elegantly. I first thought about using a `range` object, but `range` needs both a lover and an upper limit. `slice` does not. The alternative would be to return a custom object:
```python
@dataclasses.dataclass
class Range[_T]:

    min_value: _T
    max_value: _T

    def __contains__(self, x):
        return self.min_value <= x <= self.max_value

    def __str__(self):
        s = []
        if self.min_value is not None:
            s.append(f'>= {self.min_value!s}')
        if self.max_value is not None:
            s.append(f'<= {self.max_value!s}')
        return ', '.join(s)
```
The implementation of `UserIntegerOption` could be reworked to store a `Range[int]` instead than `min_value` and `max_value` directly.